File: 241211/괄호 문자열의 적합성 판단/parentheses-string.cpp
#include <iostream>
#include <stack>
using namespace std;

stack<int> stk;
string s, ans;
int main() {
    cin >> s;
    for(char c: s){
        if(c == '('){
            stk.push(1);
        } 
        else if(c == ')'){
            if(stk.empty()){
                ans = "No"; 
                break;
            }
            stk.pop();
        }
    }

    if(!stk.empty()) ans = "No";
    else ans = "Yes";
    
    cout << ans <<'\n';

    return 0;
}

File: 250221/괄호 문자열의 적합성 판단/parentheses-string.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        Stack<Integer> stack = new Stack<>();

        String s = sc.next();
        boolean check = true;
        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);

            if(c == '(') stack.push(1);
            else if(c == ')'){
                if(stack.empty()) {
                    check = false;
                    break;
                } 
                stack.pop();
            }
        }
        if(!stack.empty()) check = false;

        if(check) System.out.println("Yes");
        else System.out.println("No");
        sc.close();
    }
}